### PR TITLE
Add `RSCRoute ssr={false}` to skip initial RSC payload generation

### PR DESCRIPTION
**Pull request by ihabadham**
## Summary

Adds `ssr={false}` support to `RSCRoute`.

By default, `RSCRoute` keeps the existing `ssr={true}` behavior: the server renders the route, generates its RSC payload, embeds that payload in the HTML, and the client hydrates from the embedded payload.

When `ssr={false}`, that route instance skips initial server-side RSC payload generation. In streaming SSR, the route throws a classified server-only bailout before provider work. If it is inside a scoped `Suspense` boundary, React streams that fallback HTML and retries the route on the client.

The client retry uses the existing RSC provider path, including cache lookup, `/rsc_payload/:componentName` fetching, `ServerComponentFetchError`, and `useRSC().refetchComponent(...)` retry behavior.

Part of #3101. Follow-up #3394 adds default-provider support for deferred roots that do not manually use `wrapServerComponentRenderer`.

### API

```tsx
<RSCRoute
  componentName="HeavyWidget"
  componentProps={{ userId }}
  ssr={false}
/>
```

`ssr` defaults to `true`.

- `ssr={false}` does not call the server-side RSC payload path during initial SSR.
- Deferred routes do not get embedded `window.REACT_ON_RAILS_RSC_PAYLOADS[...]` entries.
- Suspense owns loading UI; no new fallback prop is added.
- Mixed wrapped pages work: `ssr={true}` routes can SSR while `ssr={false}` routes defer.
- Error boundary and retry behavior continue through the existing `RSCProvider` path.
- Omitting `ssr` or passing `ssr={true}` preserves the existing server-rendered path.

## Implementation notes

Rendering a user-provided `Suspense` fallback on the server requires React's server-error bailout path. Returning `null` would skip RSC payload work, but it would not activate the surrounding `Suspense` fallback. To get both behaviors, `RSCRoute ssr={false}` throws a classified server-only bailout before `useRSC`, provider cache-key generation, or RSC payload generation.

React treats that throw as recoverable when it is caught by `Suspense`: the server streams the nearest fallback and the client retries the route. React still reports that recovery path through the server stream `onError` callback and the client `hydrateRoot` recoverable-error callback, so React on Rails Pro distinguishes this intentional bailout from real render failures.

This PR:

- Adds `ssr?: boolean` to `RSCRouteProps`.
- Adds `RSCRouteSSRFalseBailoutError` to mark the intentional server-only bailout.
- Returns the bailout digest from the stream renderer so React can encode the expected retry path.
- Suppresses only the matching recoverable bailout during hydration; other recoverable errors still report normally.
- Keeps the client retry on the existing provider fetch/cache/error/retry path.

## Tests

Covered with package tests, streaming tests, docs, and dummy-app E2E for:

- skipped server payload generation;
- default `ssr={true}` behavior;
- Suspense fallback streaming;
- mixed `ssr={true}` / `ssr={false}` routes in wrapped streaming roots;
- `ServerComponentFetchError` and `useRSC().refetchComponent(...)` retry;
- no embedded payload for deferred routes;
- no broad recoverable-error suppression.

## Test plan

- `pnpm --filter react-on-rails-pro run type-check`
- `pnpm --filter react-on-rails-pro run test`
- `pnpm --filter react-on-rails run type-check`
- `pnpm --filter react-on-rails run test`
- `pnpm --filter react-on-rails run build`
- `pnpm --filter react-on-rails-pro run build`
- `ENABLE_JEST_CONSOLE=true pnpm --filter react-on-rails-pro exec jest tests/deferredRouteSsr.test.tsx --runInBand`
- `ENABLE_JEST_CONSOLE=true pnpm --filter react-on-rails-pro exec jest tests/streamServerRenderedReactComponent.test.jsx --runInBand -t 'RSCRoute ssr=false|post-SSR'`
- `cd react_on_rails_pro/spec/dummy && pnpm exec playwright test e2e-tests/rsc_route_ssr_false.spec.ts --project=chromium --config=playwright.config.ts`
- `pnpm exec prettier --check <changed files>`
- `pnpm exec eslint --no-ignore <changed source and focused test files>`
- `git diff --check`

### CODE REVIEW

File: docs/pro/react-server-components/inside-client-components.md [hunk 1]
```diff
@@ -374,7 +401,7 @@ import RSCRoute from 'react-on-rails-pro/RSCRoute';
 export default function Page({ user }) {
   return (
     <Suspense fallback={<div>Loading…</div>}>
-      <RSCRoute componentName="SlowServerComponent" componentProps={{ user }} />
+      <RSCRoute componentName="SlowServerComponent" componentProps={{ user }} ssr={false} />
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply docs/pro/react-server-components/inside-client-components.md]: Addressed in `213776466`: the general Suspense section now shows the default `RSCRoute` example first, and calls out `ssr={false}` separately as the same scoped Suspense pattern for deferred initial streaming.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 2]
```diff
@@ -88,7 +89,23 @@ const wrapServerComponentRenderer = (
...
 
     if (domNode.innerHTML) {
-      ReactDOMClient.hydrateRoot(domNode, root, { identifierPrefix: domNodeId });
+      ReactDOMClient.hydrateRoot(domNode, root, {
+        identifierPrefix: domNodeId,
+        onRecoverableError(error) {
+          const cause =
+            error instanceof Error && 'cause' in error ? (error as Error & { cause?: unknown }).cause : error;
+
+          if (isRSCRouteSSRFalseBailoutError(error) || isRSCRouteSSRFalseBailoutError(cause)) {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:110]: Addressed in `213776466`: the fallback is now `undefined`, so the handler no longer checks the same value twice when there is no `cause`.

File: packages/react-on-rails-pro/src/handleRecoverableError.client.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import { isRSCRouteSSRFalseBailoutError } from './RSCRouteSSRFalseBailoutError.ts';
+
+const getRecoverableErrorCause = (error: unknown): unknown =>
+  error instanceof Error && 'cause' in error ? (error as Error & { cause?: unknown }).cause : undefined;
+
+const handleRecoverableError = (error: unknown) => {
+  const cause = getRecoverableErrorCause(error);
+
+  if (isRSCRouteSSRFalseBailoutError(error) || isRSCRouteSSRFalseBailoutError(cause)) {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/handleRecoverableError.client.ts]: This thread is outdated for PR #3318 after the split. The shared recoverable handler belongs to the stacked default-provider PR; its current implementation keeps the conservative `cause` extraction for TypeScript compatibility while avoiding the duplicate fallback check.

File: packages/react-on-rails-pro/src/defaultRSCProviderRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * via any medium, is strictly prohibited without a valid license agreement
+ * from Shakacode LLC.
+ *
+ * For licensing terms, please see:
+ * https://github.com/shakacode/react_on_rails/blob/master/REACT-ON-RAILS-PRO-LICENSE.md
+ */
+
+import type { ReactElement } from 'react';
+import type { RailsContext } from 'react-on-rails/types';
+
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/defaultRSCProviderRegistry.ts]: This thread is no longer part of PR #3318 after the split. The default-provider registry lives in the stacked follow-up PR, so I'll keep any registry-specific follow-up there.

File: packages/react-on-rails-pro/src/RSCRoute.tsx [hunk 5]
```diff
@@ -99,4 +105,12 @@ const RSCRoute = ({ componentName, componentProps }: RSCRouteProps): ReactNode =
   );
 };
 
+const RSCRoute = ({ componentName, componentProps, ssr = true }: RSCRouteProps): ReactNode => {
+  if (!ssr && typeof window === 'undefined') {
+    throw new RSCRouteSSRFalseBailoutError(componentName);
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/RSCRoute.tsx:110]: Thanks. I validated both concerns. The `typeof window === 'undefined'` check is the established React/RoR Node-renderer pattern, so I left it unchanged. The missing scoped-Suspense case is a real UX footgun, but `RSCRoute` cannot reliably detect whether its nearest Suspense boundary is user-scoped or the wrapper root; a blanket DEV warning would also warn for valid usage. The docs now emphasize using a scoped Suspense boundary for `ssr={false}`.

File: packages/react-on-rails-pro/src/RSCRouteSSRFalseBailoutError.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  digest = RSC_ROUTE_SSR_FALSE_BAILOUT_DIGEST;
+
+  constructor(componentName: string) {
+    super(`RSCRoute "${componentName}" skipped server rendering because it was rendered with ssr={false}.`);
+    this.name = 'RSCRouteSSRFalseBailoutError';
+  }
+}
+
+export function isRSCRouteSSRFalseBailoutError(error: unknown): error is RSCRouteSSRFalseBailoutError {
+  return (
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/RSCRouteSSRFalseBailoutError.ts:40]: Addressed in `213776466`: added a short comment explaining why the digest check exists in addition to the class-instance check.

File: packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts [hunk 7]
```diff
@@ -91,10 +100,16 @@ const streamRenderReactComponent = (
           );
         },
         onError(e) {
-          reportError(convertToError(e));
+          const error = convertToError(e);
+          if (isRSCRouteSSRFalseBailoutError(error)) {
+            return error.digest;
+          }
+
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts:104]: Agreed. Leaving the code as-is here; the server-side `onError` path returns the digest intentionally so React can serialize it on the Suspense boundary, and the surrounding tests cover the classified bailout behavior.

File: packages/react-on-rails-pro/src/registerDefaultRSCProvider.client.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import { setDefaultRSCProviderFactory } from './defaultRSCProviderRegistry.ts';
+
+if (typeof window !== 'undefined') {
+  setDefaultRSCProviderFactory(({ reactElement, railsContext, domNodeId }) => {
+    const RSCProvider = createRSCProvider({
+      getServerComponent: async (args) => {
+        const { default: getReactServerComponent } = await import('./getReactServerComponent.client.ts');
+        return getReactServerComponent(domNodeId, railsContext)(args);
+      },
+    });
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/registerDefaultRSCProvider.client.tsx]: This thread is no longer part of PR #3318 after the split. The default-provider registration file lives in the stacked follow-up PR, so I'll handle any registration-specific feedback there.

File: packages/react-on-rails-pro/src/defaultRSCProviderRegistry.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  railsContext: RailsContext;
+  domNodeId: string;
+}) => ReactElement;
+
+let defaultRSCProviderFactory: DefaultRSCProviderFactory | undefined;
+
+export const setDefaultRSCProviderFactory = (factory: DefaultRSCProviderFactory) => {
+  defaultRSCProviderFactory = factory;
+};
+
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/defaultRSCProviderRegistry.ts]: This thread is no longer part of PR #3318 after the split. The default-provider registry lives in the stacked follow-up PR, so I'll keep any registry-specific follow-up there.

File: packages/react-on-rails-pro/src/handleRecoverableError.client.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+ * from Shakacode LLC.
+ *
+ * For licensing terms, please see:
+ * https://github.com/shakacode/react_on_rails/blob/master/REACT-ON-RAILS-PRO-LICENSE.md
+ */
+
+import { isRSCRouteSSRFalseBailoutError } from './RSCRouteSSRFalseBailoutError.ts';
+
+const getRecoverableErrorCause = (error: unknown): unknown =>
+  error instanceof Error && 'cause' in error ? (error as Error & { cause?: unknown }).cause : undefined;
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/handleRecoverableError.client.ts]: This thread is outdated for PR #3318 after the split. The shared recoverable handler belongs to the stacked default-provider PR; its current implementation keeps the conservative `cause` extraction for TypeScript compatibility while avoiding the duplicate fallback check.

File: packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts [hunk 11]
```diff
@@ -91,10 +100,16 @@ const streamRenderReactComponent = (
...
+          reportError(error);
+          return undefined;
         },
         onAllReady() {
-          streamingTrackers.postSSRHookTracker.notifySSREnd();
+          notifySSREndOnce();
         },
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts:115]: Agreed. Leaving the code as-is here; returning the digest from `onError` is the intentional React streaming hook that lets the client classify and suppress only this bailout.

File: packages/react-on-rails-pro/src/registerDefaultRSCProvider.client.tsx [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import { createRSCProvider } from './RSCProvider.tsx';
+import { setDefaultRSCProviderFactory } from './defaultRSCProviderRegistry.ts';
+
+if (typeof window !== 'undefined') {
+  setDefaultRSCProviderFactory(({ reactElement, railsContext, domNodeId }) => {
+    const RSCProvider = createRSCProvider({
+      getServerComponent: async (args) => {
+        const { default: getReactServerComponent } = await import('./getReactServerComponent.client.ts');
+        return getReactServerComponent(domNodeId, railsContext)(args);
+      },
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/registerDefaultRSCProvider.client.tsx]: This thread is no longer part of PR #3318 after the split. The default-provider registration file lives in the stacked follow-up PR, so I'll handle any registration-specific feedback there.

File: react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    expect(html).toContain('Mixed RSC route shell before');
+    expect(html).toContain('Server Component Title');
+    expect(html).toContain('Deferred route loading');
+    expect(html).toContain('Mixed RSC route shell after');
+    expect(html).toContain(DEFAULT_PAYLOAD_KEY);
+    expect(html).not.toContain(DEFERRED_PAYLOAD_KEY);
+    expect(html).not.toContain('Post 1');
+
+    const consoleMessages: string[] = [];
+    page.on('console', (message) => {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts:22]: This is no longer applicable to the current PR #3318 diff after the split/rebase; the current PR1 E2E file does not duplicate the bailout digest constant.

File: packages/react-on-rails-pro/src/ClientSideRenderer.ts [hunk 14]
```diff
@@ -138,10 +140,18 @@ class ComponentRenderer {
...
+            railsContext,
+            domNodeId,
+          );
           const rootOrElement = reactHydrateOrRender(
             domNode,
-            reactElementOrRouterResult as ReactElement,
+            reactElement,
             shouldHydrate,
+            shouldHydrate && wrappedByDefaultRSCProvider
+              ? { identifierPrefix: domNodeId, onRecoverableError: handleRecoverableError }
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/ClientSideRenderer.ts]: This thread is no longer part of PR #3318 after the split. The `ClientSideRenderer` changes live in the stacked default-provider PR, where this concern can be reviewed in the right context.

File: react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    await expect(page.getByText('Post 1')).toBeVisible();
+    await expect(page.getByText('Mixed RSC route shell after')).toBeVisible();
+    await expect(page.getByTestId('deferred-rsc-route-fallback')).not.toBeVisible();
+
+    const rscRequests = (await page.requests()).filter((payloadRequest) =>
+      payloadRequest.url().includes('/rsc_payload/'),
+    );
+    expect(
+      rscRequests.some((payloadRequest) => payloadRequest.url().includes('/rsc_payload/SimpleComponent')),
+    ).toBe(true);
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts:42]: Verified against this repo: the installed Playwright types expose `page.requests()`, and the existing Pro dummy E2E fixture already uses the same API. Keeping this as-is.

File: internal/planning/3101-rsc-route-ssr-false.md [hunk 16]
```diff
@@ -213,30 +220,49 @@ Example:
...
+<Footer />
+```
+
+If no nearer boundary exists in the supported wrapped streaming path, the existing root-level framework boundary
+with `fallback={null}` is the nearest boundary and catches the bailout. That preserves payload skipping, but it
+can hide the rendered root and provides no meaningful loading UI. The documented pattern is to place `Suspense`
+close to each deferred route.
+
+React still reports this documented retry path through client `onRecoverableError`. Phase 2 should not broadly
+suppress client recoverable errors; the intentional-bailout classification belongs on the server stream path.
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply internal/planning/3101-rsc-route-ssr-false.md:265]: Leaving this internal planning wording as-is for PR #3318. The file is temporary/internal and likely to be removed before merge; the PR description and user-facing docs are the review reference.

File: packages/react-on-rails-pro/src/RSCRouteSSRFalseBailoutError.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Intentional server-only bailout used by `RSCRoute ssr={false}`.
+ *
+ * React streaming SSR catches errors under Suspense boundaries and emits the nearest fallback.
+ * The stream renderer uses this classified error to avoid reporting that intentional fallback
+ * path as a real server-rendering failure.
+ */
+export class RSCRouteSSRFalseBailoutError extends Error {
+  digest = RSC_ROUTE_SSR_FALSE_BAILOUT_DIGEST;
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/RSCRouteSSRFalseBailoutError.ts]: Addressed in `213776466`: `digest` is now `readonly`.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 18]
```diff
@@ -98,7 +99,23 @@ const wrapServerComponentRenderer = (
...
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:110]: Addressed in `213776466`: the fallback is now `undefined`, so the handler no longer checks the same value twice when there is no `cause`.

File: packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts [hunk 19]
```diff
@@ -59,6 +61,13 @@ const streamRenderReactComponent = (
     pipeToTransform(errorHtmlStream);
   };
 
+  const notifySSREndOnce = () => {
+    if (didNotifySSREnd) return;
+
+    didNotifySSREnd = true;
+    streamingTrackers.postSSRHookTracker.notifySSREnd();
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts]: Superseded by the later scoped implementation. The broad `notifySSREndOnce` guard was replaced with `PostSSRHookTracker.notifySSREnd({ suppressDuplicateWarning })`, with a comment explaining the React nested-Suspense duplicate-callback case.

File: packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts [hunk 20]
```diff
@@ -59,6 +61,13 @@ const streamRenderReactComponent = (
     pipeToTransform(errorHtmlStream);
   };
 
+  const notifySSREndOnce = () => {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/streamServerRenderedReactComponent.ts]: Superseded by the later scoped implementation. The broad `notifySSREndOnce` guard was replaced with `PostSSRHookTracker.notifySSREnd({ suppressDuplicateWarning })`, and tests now verify expected bailout, unexpected nested error, and mixed bailout+real-error behavior.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:110]: Addressed in `213776466` by tightening the `cause` fallback and adding the digest comment in the classifier. The handler still checks both the current error and one `cause`, while the classifier uses the digest as the stable signal when React surfaces a recoverable boundary error instead of the original class instance.

File: packages/react-on-rails-pro/src/RSCRoute.tsx [hunk 21]
```diff
@@ -99,4 +105,12 @@ const RSCRoute = ({ componentName, componentProps }: RSCRouteProps): ReactNode =
   );
 };
 
+const RSCRoute = ({ componentName, componentProps, ssr = true }: RSCRouteProps): ReactNode => {
+  if (!ssr && typeof window === 'undefined') {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/RSCRoute.tsx:109]: Thanks. I validated this and left it unchanged: `typeof window === 'undefined'` is the established React/RoR Node-renderer pattern, and this PR targets the Node renderer context rather than edge/SSG runtimes.

File: docs/pro/react-server-components/inside-client-components.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply docs/pro/react-server-components/inside-client-components.md]: Addressed in `213776466`: the general Suspense section now shows the default `RSCRoute` example first, then explains the `ssr={false}` streaming case separately.

**Comment by ihabadham** [comment]
Address-review follow-up after `213776466`:

- Addressed the docs Suspense example by showing the default `RSCRoute` case first and documenting `ssr={false}` separately.
- Marked the bailout digest as `readonly` and documented the digest-based classifier fallback.
- Simplified recoverable-error `cause` handling so missing causes resolve to `undefined` instead of rechecking the original error.
- Investigated the broader post-SSR hook lifecycle comment. Shell/component error behavior does not reproduce as a PR1-created regression, while render-function rejection after hook registration exposes a broader pre-existing lifecycle decision. Deferring that to follow-up rather than expanding this `RSCRoute ssr={false}` PR.
- Verified `page.requests()` is available in the installed Playwright types and already used by the existing Pro dummy E2E fixture.

Resolved the review threads that were addressed, outdated after the split, or explicitly declined with the rationale above.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 22]
```diff
@@ -98,7 +99,25 @@ const wrapServerComponentRenderer = (
...
-      ReactDOMClient.hydrateRoot(domNode, root, { identifierPrefix: domNodeId });
+      ReactDOMClient.hydrateRoot(domNode, root, {
+        identifierPrefix: domNodeId,
+        onRecoverableError(error) {
+          const cause =
+            error instanceof Error && 'cause' in error
+              ? (error as Error & { cause?: unknown }).cause
+              : undefined;
+
+          if (isRSCRouteSSRFalseBailoutError(error) || isRSCRouteSSRFalseBailoutError(cause)) {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:110]: Verified against the installed React 19.2.3 source and a local hydration probe. For the actual streaming `ssr={false}` bailout path, React surfaces the digest directly on the recoverable error (`error.digest`) and `error.cause` is `undefined`. React has other recovery paths that use a single-level `{ cause }`, which is why the current one-level cause check is useful, but I did not find evidence of nested recoverable-error cause chains for this path. I’m leaving the code as-is to avoid adding speculative recursive matching.

File: react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    await deferredPayloadRequest;
+
+    await expect(page.getByTestId('mixed-rsc-route-page')).toBeVisible();
+    await expect(page.getByText('Mixed RSC route shell before')).toBeVisible();
+    await expect(page.getByText('Server Component Title')).toBeVisible();
+    await expect(page.getByText('Post 1')).toBeVisible();
+    await expect(page.getByText('Mixed RSC route shell after')).toBeVisible();
+    await expect(page.getByTestId('deferred-rsc-route-fallback')).not.toBeVisible();
+
+    const rscRequests = (await page.requests()).filter((payloadRequest) =>
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply react_on_rails_pro/spec/dummy/e2e-tests/rsc_route_ssr_false.spec.ts:37]: Verified locally: this repo uses Playwright 1.58.2, and `Page.requests()` exists in the installed Playwright types (`node_modules/playwright-core/types/types.d.ts`). The repo already uses the same API in `react_on_rails_pro/spec/dummy/e2e-tests/fixture.ts`. So this is valid for our current Playwright version and I’m leaving it unchanged.

File: packages/react-on-rails-pro/src/RSCRoute.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/RSCRoute.tsx:110]: I agree the missing scoped Suspense case is a real DX footgun, but I’m not adding a runtime warning here. `RSCRoute` cannot reliably know whether the nearest Suspense boundary is user-scoped or the wrapper’s root `fallback={null}` boundary, and the intentional bailout is suppressed during hydration by digest. The user-facing docs already call out this exact behavior and recommend adding a scoped Suspense boundary for loading UI.

File: packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx [hunk 24]
```diff
@@ -98,7 +99,25 @@ const wrapServerComponentRenderer = (
...
+          }
+
+          if (typeof globalThis.reportError === 'function') {
+            globalThis.reportError(error);
+          } else {
+            console.error(error);
+          }
+        },
+      });
     } else {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/src/wrapServerComponentRenderer/client.tsx:121]: Verified against the installed React 19.2.3 client source. React’s default `onRecoverableError` calls `reportGlobalError(error)`; it does not pass the component stack to `console.error` for recoverable errors. Our handler preserves that main behavior for non-bailout errors by using `globalThis.reportError(error)` when available and `console.error(error)` as the fallback. Since the stack-loss concern does not match the current React default behavior, I’m leaving this unchanged.

File: packages/react-on-rails-pro/tests/streamServerRenderedReactComponent.test.jsx [hunk 25]
```diff
@@ -134,6 +192,177 @@ describe('streamServerRenderedReactComponent', () => {
...
+      trace: false,
+      throwJsErrors: false,
+      railsContext: testingRailsContext,
+      generateRSCPayload,
+    });
+
+    return { renderResult, generateRSCPayload };
+  };
+
+  it('renders the nearest Suspense fallback for RSCRoute ssr=false without generating an RSC payload', async () => {
```
[the remark this answers is not in the document]
Comment (ihabadham) [reply packages/react-on-rails-pro/tests/streamServerRenderedReactComponent.test.jsx:289]: This edge case is documented: without a nearby scoped Suspense boundary, the wrapper’s root `fallback={null}` boundary may leave the deferred area empty until the client resolves the route. The current tests focus on the supported/recommended scoped-Suspense behavior plus bailout classification and error handling. I’m not adding another test for the intentionally discouraged root-fallback-null shape in this PR.

**Comment by ihabadham** [comment]
## Address-review summary

Scan scope: full PR history; no prior summary marker.

### Mattered
- None. No correctness/blocking issues were accepted from the latest review pass.

### Optional
- Declined recursive recoverable-error cause-chain matching after verifying the current React 19.2.3 bailout path carries the digest directly on the recoverable error and does not use nested causes.
- Declined changing non-bailout recoverable-error reporting after verifying React 19.2.3's default recoverable-error handler calls `reportGlobalError(error)` and does not include component-stack context.
- Declined adding a no-user-Suspense streaming test; the documented supported/recommended path remains scoped Suspense around `RSCRoute ssr={false}`.

### Skipped
- Skipped the `page.requests()` change request after verifying Playwright 1.58.2 includes `Page.requests()` and this repo already uses it in the E2E fixture.
- Skipped a runtime warning/dev hint for missing scoped Suspense because `RSCRoute` cannot reliably distinguish a user-scoped Suspense boundary from the wrapper root `fallback={null}` boundary; docs already describe the behavior.
- Skipped test-style feedback around `jest.doMock`/`require()` because that shape is intentional for module-isolation around the client wrapper side-effect import.

Follow-up issue: None.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.
